File: GoLangChat/Server/Chat/Packet.go
// coded by Jongwhan Lee   (leejw51@gmail.com)
// Session
package main

//	"github.com/golang/protobuf/proto"

//"fmt"
//	"net"1
//	"time"
//	"os"
//	"bytes"
//	"encoding/binary"
//	"sync"

type PacketType int

const (
	PacketBasicMessage = 0 // google protobuf
	PacketJsonMessage  = 1 // just json
)

type Packet struct {
	head PacketHead
	body []byte
}

func (packet *Packet) pack(src []byte) {
	buf := src
	packet.body = buf
	packet.head.length = int64(len(packet.body))
}

func (packet *Packet) unpack() (dst []byte) {
	buf := packet.body
	return buf[0:]
}

type Protocol interface {
	onReceiveBasicMessage(packet *Packet) // google protobuf
	onReceiveJsonMessage(packet *Packet)  // json
}

func parse(protocol Protocol, packet *Packet) {
	switch packet.head.id {
	case PacketBasicMessage:
		protocol.onReceiveBasicMessage(packet)

	case PacketJsonMessage:
		protocol.onReceiveJsonMessage(packet)
	}
}

File: GoLangChat/Server/ChatConsole/VirtualChat/VirtualChat/Packet.cs
/*
 coded by Jongwhan Lee

leejw51@gmail.com

@2016 Bumblebee
 */
using System;
using System.Threading;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;


namespace VirtualChat
{
	public class Packet
	{
		public Int64 id;
		public Int64 length;
		// body length
		public byte[] body;

		public void packJson(string json)
		{
			
			this.id = 1;
			this.body = System.Text.Encoding.UTF8.GetBytes (json);
			this.length = this.body.Length;
		}
	}

	public class WorkMessage
	{
		public Packet packet;
	}

}

